### ISSUE TITLE
COMMS - Read exact message type with ProtocolStack

### ISSUE THREAD

**Issue by akowalew** [arobenko/comms_champion#4]
Hello!

I'm a big fan of this library. I've successfully used it in my project. However, I ran into one problem: how to read only one, known at compile time, message type, using `ProtocolStack`? In the documentation it is written, that messages can be handled using `dispatch` method of `Message`. It is very useful, I know, because of polymorphic behaviour, but brings unnecessary overhead in my case. 

In the protocol I'm working at, I can send to server some `Command`, and then it must send an `Answer`. I know at compile time, which type of `Answer` it should be, because It comes from this protocol specification. It would be useful, in my opinion, to write code like this:

```c++
ConcreteAnswerMsg concreteAnswerMsg;
const auto errorStatus = protocolStack.tryRead(inputBuffer, concreteAnswerMsg);
if(errorStatus == Success)
{
    // good, input buffer contains message needed by us
    // read finished successfully, so call the success callback
    assert(concreteAnswerMsg.field_data().valid());
    const auto& data = concreteAnswerMsg.field_data().value();
    callback(data);
}
```

instead of writing code, like at the moment:
```c++
MessagePtr msgPtr;
...
const auto errorStatus = protocolStack.read(msgPtr, ...);
if(errorStatus == Success)
{
    assert(msgPtr);
    msgPtr->dispatch(handler);
    // message handler has to accept only `ConcreteAnswerMsg` type (by calling the success callback), 
    // and reject all others(by calling the error callback)
}
```

**Comment by arobenko** [comment]
@akowalew, Thanks for the suggestion and new idea, I'll consider implementing something like that in the future versions of the library. 

When speaking about "unnecessary overhead", do you mean performance penalty by executing indirect call of virtual function or unnecessary bloat of v-table (and as the result bloat the binary size) with extra virtual function?

Note, that the library was designed to be flexible enough, so the developer could substitute bits and pieces to adjust the default behavior. You can eliminate polymorphic dispatch functionality by implementing extra logic yourself. I'd recommend using `readFieldsCached()` member function of the protocol stack, instead of `read()`, and then check the value of message ID field inside.
```
ProtocolStack::AllFields cachedTransportFields;
auto es = protocolStack.readFieldsCached(cachedTransportFields, msgPtr, ...);
if (es != comms::ErrorStatus::Succes) { /* handle error */ return }

auto& msgIdField = std::get<idxOfMsgIdField>(cachedTransportFields);
auto msgIdValue = msgIdField.value();
if (msgIdValue == IdOfAnswer) {
    auto* answerMsg = static_cast<ConcreteAnswerMsg*>(msgPtr.get());
    handler.handle(*answerMsg); // call appropriate handling function directly without polymorphic dispatch
}
```
**WARNING:** Caching of the transport fields will also cache the whole payload (handled by `comms::protocol::MsgDataLayer`) as `comms::field::ArrayList` with `std::vector<std::uint8_t>` as inner storage by default. In most cases it's **NOT** what you want. You'll have to change the default template argument to `comms::protocol::MsgDataLayer` (when defining your protocol stack type) to be the same `comms::field::ArrayList`, but which also receives extra `comms::option::OrigDataView` option. It will make the inner storage to be `comms::util::ArrayView` (instead of `std::vector`), which is just a view on original data without copying.

**Comment by akowalew** [comment]
Thank you for your answer.

Mainly I'm thinking about performance penalty. Code size at the moment does not matter, because I have a lot o flash memory in microcontroller. ;)

After all I've written code similar to yours - by checking Message ID at the end - but in my opinion, there is still unnecessary overhead, and dangerous casting of pointer. We are still reading message in polymorphic way, even though we are expecting only one, known at compile time, message type. It would be nice to have a function, which reads only concrete message type and rejects, if input buffer contains anything else. :)

**Comment by arobenko** [comment]
Yes, I agree. I will definitely consider introducing such a feature in one of the future versions.

**Comment by arobenko** [comment]
I have implemented the requested feature and it is currently available on "**develop**" branch. It will be merged to master the next release (hopefully end of this month).

The resolution is that every **read()** and **readFieldsCached()** member function of every available protocol layer can receive in its parameters either reference to a smart pointer to allocated message (**ProtocolStack::MsgPtr**) or reference the actual message itself (which extends **comms::MessageBase**). There is inner compile time template meta-programming magic to determine type of the passed parameter and act accordingly. 

In case of passing a reference to actual message object and reading unexpected message ID value, **comms::ErrorStatus::InvalidMsgId** will be returned. If the read message ID is equal to expected one, the read operation proceeds further to the next layer. The final data layer (**comms::protocol::MsgDataLayer**) will invoke non-virtual **doRead()** member function of the message object directly instead of via call to polymorphic **read()**, which eliminates a need to any virtual function when defining an interface for input messages.

Please try it and tell me whether it works for you. Note, that the support for dual nature of "**msg**" parameter is supported in all the provided layers. However, if you had to implement custom one, there may be a need to do some extra compile time checks and act accordingly in your "**doRead()**" member function.

**Comment by akowalew** [comment]
Works great. It is not calling any virtual function, and none of them is placed in final application.

Thank you very much!

**Comment by arobenko** [comment]
Good. I'm in the process in rewriting project documentation and tutorials. Once I finish it the code is going to be released to master branch (hopefully until end of the month).

**Comment by arobenko** [comment]
Officially released in v0.24
